### ISSUE TITLE
IsDaylightSavingTime returning False when it should be returning True

### ISSUE THREAD

**Issue by carlblanchard** [dotnet/runtime#84563]
### Description

I've been working with localising dates/times in my project.
When evaluating if a date is inside daylight saving time for the UK, I am getting a false value when I should be getting a true value.
Here is the official documentation - on the 29th October 2023 @ 02:00:00 the time will become 01:00:00 

![image](https://user-images.githubusercontent.com/18427214/230905412-51ef9142-1080-4975-8058-7ba3ec60f9eb.png)

I wrote a C# Fiddle which proves this problem exists  https://dotnetfiddle.net/QgXoI6

```Csharp
using System;
using System.Globalization;

public class Program
{
	public static void Main()
	{ 
		var timeZone = TimeZoneInfo.FindSystemTimeZoneById("Europe/London");

		Console.WriteLine(timeZone.DaylightName);
		Console.WriteLine(timeZone.SupportsDaylightSavingTime);
		
		
		var insideDateAsTicks = new DateTime(2023,10,29,01,59,59, DateTimeKind.Local).Ticks;
		var outsideDateAsTicks = new DateTime(2023,10,29,02,00,01, DateTimeKind.Local).Ticks;
		
		var insideDSTUK = new DateTime(insideDateAsTicks); //29/10/2023 01:59:59
		var outsideDSTUK = new DateTime(outsideDateAsTicks); //29/10/2023 02:00:01
		
		var culture = new CultureInfo("en-GB");
		
		//Expected result for IsDaylightSavingTime would be true as the date & time is inside UK's DST (THIS IS NOT CORRECT)
		Console.WriteLine($"Inside Date Time {insideDSTUK.ToString(culture)} value of IsDaylightSavingTime {timeZone.IsDaylightSavingTime(insideDSTUK)}");
		
		//Expected result for IsDaylightSavingTime would be false as the date & time is outside UK's DST (This is correct)
		Console.WriteLine($"Outside Date Time {outsideDSTUK.ToString(culture)} value of IsDaylightSavingTime {timeZone.IsDaylightSavingTime(insideDSTUK)}");

	}
}
```
From this fiddle, my expected results should be TRUE/FALSE however you get FALSE/FALSE 
when changing the date to the 28th oct on line 14 I get the result TRUE/TRUE when I am expecting TRUE/FALSE.

### Reproduction Steps

Run example code and change dates to be inside or outside of British Daylight Saving Time

Change Line 14 with the following values 

//One second before the BST ends 
var insideDateAsTicks = new DateTime(2023,10,29,01,59,59, DateTimeKind.Local).Ticks;   | Expected Return TRUE actual FALSE

//One entire day before BST ends 
var insideDateAsTicks = new DateTime(2023,10,28,13,59,59, DateTimeKind.Local).Ticks;  | Expected Return TRUE actual TRUE

### Expected behavior

For a date that falls inside British Summer Time then the function TimeZoneInfo.IsDaylightSavingTime function should return a true result. 

### Actual behavior

When checking values on the last day of ( British Summer Time /  Daylight Saving Time ) the function returns the wrong result.

**Comment by tarekgh** [comment]
You may use [IsAmbiguousTime(DateTime)](https://learn.microsoft.com/en-us/dotnet/api/system.timezoneinfo.isambiguoustime?view=net-7.0#system-timezoneinfo-isambiguoustime(system-datetime)) to detect if the time is ambiguous too if this can help. 

closing per @Clockwork-Muse comment.

**Comment by carlblanchard** [comment]
@Clockwork-Muse  So you are saying changing lines 14 & 15 to Utc should resolve this issue?

Returns the same False/False result?
![image](https://user-images.githubusercontent.com/18427214/230948102-472f0e6a-bc61-4f26-801e-e23b29c79b5d.png)

**Comment by tarekgh** [comment]
@carlblanchard 

> So you are saying changing lines 14 & 15 to Utc should resolve this issue?

No, if you are using local time you should use [IsAmbiguousTime(DateTime)](https://learn.microsoft.com/en-us/dotnet/api/system.timezoneinfo.isambiguoustime?view=net-7.0#system-timezoneinfo-isambiguoustime(system-datetime)) to know if the time is IsAmbiguousTime and decide to do whatever you want. If you are using Utc time (which is not the case in your code), you can call IsDaylightSavingTime on that.

**Comment by carlblanchard** [comment]
@tarekgh I have updated the sample to use UTC, still has the same issue.

**Comment by tarekgh** [comment]
> I have updated the sample to use UTC, still has the same issue.

Yes, this is expected. It is wrong to `UTC` with the local time. as I mentioned before you need to call [IsAmbiguousTime(DateTime)](https://learn.microsoft.com/en-us/dotnet/api/system.timezoneinfo.isambiguoustime?view=net-7.0#system-timezoneinfo-isambiguoustime(system-datetime)) instead. If this return `true`, you'll know this time occur twice in that day. One with daylight saving on and one when the daylight saving became off.

**Comment by carlblanchard** [comment]
Sorry, I don't quite follow. The date 29 Oct 2023 01:59:59 is inside DST in the UK.
Surely, if I pass that date into the function IsDaylightSavingTime it should return true right, after all that date is inside DST.
If I change the day part to 28 from 29 then it returns True.

**Comment by Clockwork-Muse** [comment]
> I have updated the sample to use UTC, still has the same issue.

Well no, what you've actually done is given yourself a different issue.  That time in UTC is not the same instant in the local London time zone.  You can't just put in UTC and change nothing else and expect it to work.

(In the case of London specifically it only fails because the zone you're trying to use is an hour ahead until after the transition.  But that is not the general case)

Note that you'd get the correct answer if you were using `DateTimeOffset` instead... however...

If you're doing **any** complicated work with date/time in C#, you really ought to be using [NodaTime](https://nodatime.org/), as many of the things you want to do are non-trivial in the base APIs.  For example, here you'd have to manually adjust the converted instance, which has more edge cases than you might expect (the date DST occurs, the time DST occurs, the amount of the offset...).

**Comment by carlblanchard** [comment]
Yeah, I see the conundrum there. However, my test & example is pretty simple. 1 second before the cut of time should return true, and 1 second after should return false. According to the documentation for the function

Surely passing in a date to a function starting with "Is" should return a Yes or No response? regardless how many times it occurs during the day... 

  https://learn.microsoft.com/en-us/dotnet/api/system.datetime.isdaylightsavingtime?view=net-6.0
![image](https://user-images.githubusercontent.com/18427214/230960036-efcd5d10-675c-41fc-ad45-96b8f37f390f.png)



Even embedding the example ResolveAmbiguousTime from https://learn.microsoft.com/en-us/dotnet/standard/datetime/resolve-ambiguous-times into a new C# Fiddle yields obscure results https://dotnetfiddle.net/USnXwb

Okay, rather than going back and forth clearly you know more than me in this subject, so what about helping me out with an example of code perhaps a c# fiddle demonstrating the function ```IsDayLightSavingTime``` returning true for the date of 29th Oct 2023 01:59:59 for the TimeZoneId "Europe/London"

**Comment by tarekgh** [comment]
> Surely passing in a date to a function starting with "Is" should return a Yes or No response? regardless how many times it occurs during the day...

https://learn.microsoft.com/en-us/dotnet/api/system.timezoneinfo.isdaylightsavingtime?view=net-6.0#system-timezoneinfo-isdaylightsavingtime(system-datetime)
```
If the dateTime parameter specifies an ambiguous time in the current object's time zone, the [TimeZoneInfo.IsDaylightSavingTime](https://learn.microsoft.com/en-us/dotnet/api/system.timezoneinfo.isdaylightsavingtime?view=net-6.0) method interprets dateTime as standard time and returns false if its [Kind](https://learn.microsoft.com/en-us/dotnet/api/system.datetime.kind?view=net-6.0) property is [DateTimeKind.Local](https://learn.microsoft.com/en-us/dotnet/api/system.datetimekind?view=net-6.0#system-datetimekind-local) or [DateTimeKind.Unspecified](https://learn.microsoft.com/en-us/dotnet/api/system.datetimekind?view=net-6.0#system-datetimekind-unspecified). If the [Kind](https://learn.microsoft.com/en-us/dotnet/api/system.datetime.kind?view=net-6.0) property is [DateTimeKind.Utc](https://learn.microsoft.com/en-us/dotnet/api/system.datetimekind?view=net-6.0#system-datetimekind-utc), this method will select the correct ambiguous time and indicate whether it is a daylight saving time.
```

> what about helping me out with an example of code perhaps a c# fiddle demonstrating the function IsDayLightSavingTime returning true for the date of 29th Oct 2023 01:59:59 for the TimeZoneId "Europe/London"

I added the following to your fiddle:

```C#
private static bool IsDst(TimeZoneInfo tzi, DateTime dateTime) => tzi.IsDaylightSavingTime(dateTime) || tzi.IsAmbiguousTime(dateTime);
````

then calling 

```C#
        Console.WriteLine($"*** {insideDSTUK} is DST: {IsDst(timeZone, insideDSTUK)}");
        Console.WriteLine($"*** {outsideDSTUK} is DST: {IsDst(timeZone, outsideDSTUK)}");
```

produces 

```
*** 10/29/2023 01:59:59 is DST: True
*** 10/29/2023 02:00:01 is DST: False
```

**Comment by carlblanchard** [comment]
@tarekgh You get a gold star :)
![image](https://user-images.githubusercontent.com/18427214/230965161-abf5b836-cc3e-42a8-ad5e-cf7a90185054.png)

You've given me a headache (in a good way) with that one-liner :) Today I learned something new. 

Thank you for your help @tarekgh  & @Clockwork-Muse
